File: src/main/java/leets/weeth/global/auth/authentication/CustomAccessDeniedHandler.java
package leets.weeth.global.auth.authentication;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import leets.weeth.global.common.response.CommonResponse;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.web.access.AccessDeniedHandler;
import org.springframework.stereotype.Component;

import java.io.IOException;

@Slf4j
@Component
public class CustomAccessDeniedHandler implements AccessDeniedHandler {

    @Override
    public void handle(HttpServletRequest request, HttpServletResponse response, AccessDeniedException accessDeniedException) throws IOException, ServletException {
        setResponse(response);
        log.error("ExceptionClass: {}, Message: {}", accessDeniedException.getClass().getSimpleName(), accessDeniedException.getMessage());
    }

    private void setResponse(HttpServletResponse response) throws IOException {
        response.setStatus(HttpServletResponse.SC_FORBIDDEN);
        response.setContentType("application/json");
        response.setCharacterEncoding("UTF-8");

        String message = new ObjectMapper().writeValueAsString(CommonResponse.createFailure(ErrorMessage.FORBIDDEN.getCode(), ErrorMessage.FORBIDDEN.getMessage()));
        response.getWriter().write(message);
    }
}

File: src/main/kotlin/com/weeth/global/auth/authentication/CustomAccessDeniedHandler.kt
package com.weeth.global.auth.authentication

import com.fasterxml.jackson.databind.ObjectMapper
import com.weeth.global.auth.jwt.application.exception.JwtErrorCode
import com.weeth.global.common.response.CommonResponse
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.access.AccessDeniedHandler
import org.springframework.stereotype.Component

@Component
class CustomAccessDeniedHandler(
    private val objectMapper: ObjectMapper,
) : AccessDeniedHandler {
    private val errorLog = LoggerFactory.getLogger("ERROR_LOG")

    override fun handle(
        request: HttpServletRequest,
        response: HttpServletResponse,
        accessDeniedException: AccessDeniedException,
    ) {
        if (isTemporaryUser()) {
            setRegistrationIncompleteResponse(response)
        } else {
            setForbiddenResponse(response)
        }

        try {
            MDC.put("status", response.status.toString())
            MDC.put("errorType", accessDeniedException::class.simpleName ?: "AccessDeniedException")
            MDC.put("errorMessage", accessDeniedException.message ?: ErrorMessage.FORBIDDEN.message)
            errorLog.warn("Access Denied")
        } finally {
            MDC.remove("status")
            MDC.remove("errorType")
            MDC.remove("errorMessage")
        }
    }

    private fun isTemporaryUser(): Boolean =
        SecurityContextHolder
            .getContext()
            .authentication
            ?.authorities
            ?.any { it.authority == "ROLE_TEMPORARY" }
            ?: false

    private fun setRegistrationIncompleteResponse(response: HttpServletResponse) {
        val errorCode = JwtErrorCode.REGISTRATION_INCOMPLETE
        response.status = errorCode.status.value()
        response.contentType = "application/json"
        response.characterEncoding = "UTF-8"

        val body =
            objectMapper.writeValueAsString(
                CommonResponse.error(errorCode),
            )
        response.writer.write(body)
    }

    private fun setForbiddenResponse(response: HttpServletResponse) {
        response.status = HttpServletResponse.SC_FORBIDDEN
        response.contentType = "application/json"
        response.characterEncoding = "UTF-8"

        val body =
            objectMapper.writeValueAsString(
                CommonResponse.createFailure(
                    ErrorMessage.FORBIDDEN.code,
                    ErrorMessage.FORBIDDEN.message,
                ),
            )
        response.writer.write(body)
    }
}
